src/System.IO.Packaging/src/System/IO/Packaging/ContentType.cs

```diff
@@ -0,0 +1,727 @@
+                else
+                {
+                    return x.AreTypeAndSubTypeEqual(y, true);
+                }
+            }
+
+            /// <summary>
+            /// We lower case the results of ToString() because it returns the original
+            /// casing passed into the constructor.  ContentTypes that are equal (which
+            /// ignores casing) must have the same hash code.
+            /// </summary>
+            public int GetHashCode(ContentType obj)
+            {
+                return obj._type.ToUpperInvariant().GetHashCode() ^ obj._subType.ToUpperInvariant().GetHashCode();
+            }
+        }
+        #endregion Nested Classes
+
+        //------------------------------------------------------
+        //
+        //  Private Methods
+        //
+        //------------------------------------------------------
+
+        #region Private Methods
+
+
+        /// <summary>
+        /// This method validates if the content type string has
+        /// valid CR-LF characters. Specifically we test if '\r' is 
+        /// accompanied by a '\n' in the string, else its an error.
+        /// </summary>
+        /// <param name="contentType"></param>
+        private static void ValidateCarriageReturns(string contentType)
+        {
+            Debug.Assert(!IsLinearWhiteSpaceChar(contentType[0]) && !IsLinearWhiteSpaceChar(contentType[contentType.Length - 1]));
+
+            //Prior to calling this method we have already checked that first and last
+            //character of the content type are not Linear White Spaces. So its safe to
+            //assume that the index will be greater than 0 and less that length-2.
+
+            int index = contentType.IndexOf(s_linearWhiteSpaceChars[2]);
+
+            while (index != -1)
+            {
+                if (contentType[index - 1] == s_linearWhiteSpaceChars[1] || contentType[index + 1] == s_linearWhiteSpaceChars[1])
+                {
+                    index = contentType.IndexOf(s_linearWhiteSpaceChars[2], ++index);
+                }
+                else
+                    throw new ArgumentException(SR.InvalidLinearWhiteSpaceCharacter);
+            }
+        }
+
+        /// <summary>
+        /// Parses the type ans subType tokens from the string. 
+        /// Also verifies if the Tokens are valid as per the grammar.
+        /// </summary>
+        /// <param name="typeAndSubType">substring that has the type and subType of the content type</param>
+        /// <exception cref="ArgumentException">If the typeAndSubType parameter does not have the "/" character</exception>
+        private void ParseTypeAndSubType(string typeAndSubType)
+        {
+            //okay to trim at this point the end of the string as Linear White Spaces(LWS) chars are allowed here.
+            typeAndSubType = typeAndSubType.TrimEnd(s_linearWhiteSpaceChars);
+
+            string[] splitBasedOnForwardSlash = typeAndSubType.Split(s_forwardSlashSeparator);
```

Comment:
There's a lot of string manipulation / allocation / etc. happening in these various helper functions.  Doesn't need to be fixed now, as it's not clear that it's a problem.  But it'd be interesting to do some perf testing / profiling and see what pops as hot spots that we'll potentially want to fix.